sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,303 @@
+}
+
+func (lc *locationCache) markEndptUnavailableForWrite(endpoint url.URL) error {
+	return lc.markEndptUnavailable(endpoint, write)
+}
+
+func (lc *locationCache) markEndptUnavailable(endpoint url.URL, op opType) error {
+	now := time.Now()
+	lc.mapMutex.Lock()
+	if info, ok := lc.locationUnavailabilityInfoMap[endpoint]; ok {
+		info.lastCheckTime = now
+		info.unavailableOps |= op
+		lc.locationUnavailabilityInfoMap[endpoint] = info
+	} else {
+		info = locationUnavailabilityInfo{
+			lastCheckTime:  now,
+			unavailableOps: op,
+		}
+		lc.locationUnavailabilityInfoMap[endpoint] = info
+	}
+	lc.mapMutex.Unlock()
+	err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+	return err
+}
+
+func (lc *locationCache) dbAcctRead(dbAcct AcctProperties) error {
+	return lc.update(dbAcct.writeRegions, dbAcct.readRegions, nil, dbAcct.enableMultipleWriteLocations)
+}
+
+func (lc *locationCache) refreshStaleEndpts() {
+	lc.mapMutex.Lock()
+	for endpoint, info := range lc.locationUnavailabilityInfoMap {
+		t := time.Since(info.lastCheckTime)
+		if t > lc.unavailableLocationExpirationTime {
+			delete(lc.locationUnavailabilityInfoMap, endpoint)
+		}
+	}
+	lc.mapMutex.Unlock()
+}
+
+func (lc *locationCache) isEndptUnavailable(endpoint url.URL, ops opType) bool {
+	lc.mapMutex.RLock()
+	info, ok := lc.locationUnavailabilityInfoMap[endpoint]
+	lc.mapMutex.RUnlock()
+	if ops == none || !ok || ops&info.unavailableOps != ops {
+		return false
+	}
+	lc.rwMutex.RLock()
+	defer lc.rwMutex.RUnlock()
+	return time.Since(info.lastCheckTime) < lc.unavailableLocationExpirationTime
+}
+
+func (lc *locationCache) getPrefAvailableEndpts(endptsByLoc map[string]url.URL, locs []string, availOps opType, fallbackEndpt url.URL) []url.URL {
+	endpts := make([]url.URL, 0)
+	lc.rwMutex.RLock()
+	if lc.enableEndptDiscovery {
+		if lc.canUseMultipleWriteLocs() || availOps&read != 0 {
+			unavailEndpts := make([]url.URL, 0)
+			unavailEndpts = append(unavailEndpts, fallbackEndpt)
+			for _, loc := range lc.locationInfo.prefLocations {
+				if endpt, ok := endptsByLoc[loc]; ok && endpt != fallbackEndpt {
+					if lc.isEndptUnavailable(endpt, availOps) {
+						unavailEndpts = append(unavailEndpts, endpt)
+					} else {
+						endpts = append(endpts, endpt)
+					}
+				}
+			}
+			endpts = append(endpts, unavailEndpts...)
+		} else {
+			for _, loc := range locs {
+				if endpt, ok := endptsByLoc[loc]; ok && loc != "" {
+					endpts = append(endpts, endpt)
+				}
+			}
+		}
+	}
+	lc.rwMutex.RUnlock()
+	if len(endpts) == 0 {
+		endpts = append(endpts, fallbackEndpt)
+	}
+	return endpts
+}
+
+func getEndptsByLocation(locs []AcctRegion) (map[string]url.URL, []string, error) {
+	endptsByLoc := make(map[string]url.URL)
+	parsedLocs := make([]string, 0)
+	for _, loc := range locs {
+		endpt, err := url.Parse(loc.endpoint)
+		if err != nil {
+			return nil, nil, err
+		}
+		if loc.name != "" {
+			endptsByLoc[loc.name] = *endpt
+			parsedLocs = append(parsedLocs, loc.name)
+		}
```

Comment:
Probably needs an else clause with a comment like a TODO: Log ? I know that we cannot do the logging now but at least leave a TODO in all the places where we might want to